### PR TITLE
feat(api): add `.tables` accessor to `BaseBackend`

### PR DESCRIPTION
**Pull request by jcrist**
This adds a `.tables` access to the `BaseBackend`. The main motivation here is tab-completion for table access when working with multiple tables in an interactive session. Tab completion works via both attribute and index access:

image

~Still needs tests/docs, just pushing this up for now to see if people like the idea.~

### CODE REVIEW

**Comment by gforsyth** [comment]
> * Putting this on the `.table` attribute (so it's a function and accessor at the same time). 

This was what I initially thought of when you proposes this but I prefer the plural accessor

File: ibis/backends/base/__init__.py [hunk 1]
```diff
@@ -370,6 +426,20 @@ def exists_table(self, name: str, database: str | None = None) -> bool:
...
+    @cached_property
+    def tables(self):
+        """An accessor for tables in the database.
+
+        Tables may be accessed by name using either index or attribute access:
+
+        Examples
+        --------
+        >>> con = ibis.sqlite.connect("example.db")
+        >>> people = ibis.tables['people']  # access via index
```
Comment (cpcloud) [inline ibis/backends/base/__init__.py:438]:
```suggestion
        >>> people = con.tables['people']  # access via index
```

**Review by cpcloud — APPROVED** [review]
My 2c: I think `table` versus `tables` is fine. At some point we may want to make the `__repr__` of this thing look nice.

Couple comments on the docstrings, otherwise looks good.

File: ibis/backends/base/__init__.py [hunk 2]
```diff
@@ -147,6 +157,52 @@ def list_tables(self, like=None):
...
+
+    def __init__(self, backend: BaseBackend):
+        self._backend = backend
+
+    def __getitem__(self, name) -> ir.Table:
+        return self._backend.table(name)
+
+    def __getattr__(self, name) -> ir.Table:
+        if name.startswith("_"):
+            raise AttributeError(name)
```
[the remark this answers is not in the document]
Comment (jcrist) [reply ibis/backends/base/__init__.py:180]: We at least need to efficiently filter out `__getattr__` hits for missing magic methods, I figured might as well also do it for private methods. I think forcing users to use `__getitem__` access for leading-underscore tables is fine.

**Review by cpcloud — APPROVED** [review]
LGTM. This will help usability quite a bit.

**Pull request merged**